Lib9c/Action/CombinationEquipment.cs

```diff
@@ -20,262 +17,358 @@
                     .SetState(worldInformationAddress, MarkChanged)
                     .SetState(questListAddress, MarkChanged)
-                    .MarkBalanceChanged(GoldCurrencyMock, ctx.Signer, BlacksmithAddress);
+                    .MarkBalanceChanged(GoldCurrencyMock, context.Signer, BlacksmithAddress);
             }
 
-            var addressesHex = GetSignerAndOtherAddressesHex(context, AvatarAddress);
+            var addressesHex = GetSignerAndOtherAddressesHex(context, avatarAddress);
 
-            if (!states.TryGetAgentAvatarStatesV2(ctx.Signer, AvatarAddress, out var agentState,
+            if (!states.TryGetAgentAvatarStatesV2(context.Signer, avatarAddress, out var agentState,
                 out var avatarState))
             {
-                throw new FailedLoadStateException($"{addressesHex}Aborted as the avatar state of the signer was failed to load.");
+                throw new FailedLoadStateException(
+                    $"{addressesHex}Aborted as the avatar state of the signer was failed to load.");
+            }
+
+            // Validate Required Cleared Stage
+            if (!avatarState.worldInformation.IsStageCleared(
+                GameConfig.RequireClearedStageLevel.CombinationEquipmentAction))
+            {
+                avatarState.worldInformation.TryGetLastClearedStageId(out var current);
+                throw new NotEnoughClearedStageLevelException(
+                    addressesHex,
+                    GameConfig.RequireClearedStageLevel.CombinationEquipmentAction,
+                    current);
             }
+            // ~Validate Required Cleared Stage
 
-            var slotState = states.GetCombinationSlotState(AvatarAddress, SlotIndex);
+            // Validate SlotIndex
+            var slotState = states.GetCombinationSlotState(avatarAddress, slotIndex);
             if (slotState is null)
             {
-                throw new FailedLoadStateException($"{addressesHex}Aborted as the slot state is failed to load");
+                throw new FailedLoadStateException(
+                    $"{addressesHex}Aborted as the slot state is failed to load: # {slotIndex}");
             }
 
-            if (!slotState.Validate(avatarState, ctx.BlockIndex))
+            if (!slotState.Validate(avatarState, context.BlockIndex))
             {
                 throw new CombinationSlotUnlockException(
-                    $"{addressesHex}Aborted as the slot state is invalid: {slotState} @ {SlotIndex}");
+                    $"{addressesHex}Aborted as the slot state is invalid: {slotState} @ {slotIndex}");
             }
+            // ~Validate SlotIndex
 
-            var recipeSheet = states.GetSheet<EquipmentItemRecipeSheet>();
-            var materialSheet = states.GetSheet<MaterialItemSheet>();
-            var materials = new Dictionary<Material, int>();
+            // Validate Work
+            var costActionPoint = 0;
+            var costNCG = 0L;
```

Comment:
i think use BigIteger(or decimal) is better than float type.

Reply:
Where is the float type? `costNCG` is long type.